mono/metadata/threads.c

```diff
@@ -1639,10 +1640,12 @@ ves_icall_System_Threading_Thread_Join_internal(MonoThread *this_obj, int ms)
 	mono_thread_set_state (cur_thread, ThreadState_WaitSleepJoin);
 
 	MONO_ENTER_GC_SAFE;
-	ret=mono_thread_info_wait_one_handle (handle, ms, TRUE);
+	ret=mono_wait_uninterrupted (cur_thread, 1, &handle, FALSE, ms, &error);
```

Comment:
That will not work, and I am pretty surprised it didn't SISEGV.
`mono_thread_info_wait_one_handle` is for a `MonoThreadHandle*` and `mono_wait_uninterrupted` is for a `gpointer` (aka `MonoW32HandleBase*` but that's a w32handle implementation detail for unix).
If you want it to not return due to a APC, you should check the `ret` value to be `MONO_THREAD_INFO_WAIT_RET_ALERTED` and react accordingly. You would have to check for interruption in this case (look at https://github.com/Unity-Technologies/mono/blob/3e64049842c487346f72416b57a9d05293566432/mono/metadata/threads.c#L1693)